### PR TITLE
RSDK-4053 add Camera::GetImages

### PR DESCRIPTION
**Pull request by bhaney**

### CODE REVIEW

**Comment by bhaney** [comment]
Still need to add tests - will do that tomorrow morning

**Comment by bhaney** [comment]
@cheukt FYI

File: src/viam/sdk/components/camera/server.cpp [hunk 1]
```diff
@@ -60,11 +63,63 @@ ::grpc::Status CameraServer::GetImage(::grpc::ServerContext* context,
...
+        ::viam::component::camera::v1::Image proto_image;
+        const std::string img_string = bytes_to_string(img.bytes);
+        proto_image.set_source_name(img.source_name);
+        proto_image.set_format(MIME_string_to_format(img.mime_type));
+        proto_image.set_image(img_string);
+        *response->mutable_images()->Add() = std::move(proto_image);
+    }
+    viam::common::v1::ResponseMetadata* resp_metadata = response->mutable_response_metadata();
+    google::protobuf::Timestamp ts = time_pt_to_timestamp(image_coll.captured_at);
+    resp_metadata->mutable_captured_at()->CopyFrom(ts);
```
Comment (bhaney) [inline src/viam/sdk/components/camera/server.cpp:120]: strangely, I could do `resp_metadata->set_captured_at(ts);`, it said the method didn't exist. So I did this round-about method of copying the timestamp into the response.

File: src/viam/sdk/components/camera/camera.hpp [hunk 2]
```diff
@@ -87,6 +87,14 @@ class Camera : public Component {
     struct raw_image {
         std::string mime_type;
         std::vector<unsigned char> bytes;
+        std::string source_name;
```
Comment (bhaney) [inline src/viam/sdk/components/camera/camera.hpp:90]: modified raw_images to add the source_name, it should not be a breaking change.

File: src/viam/sdk/components/camera/server.cpp [hunk 3]
```diff
@@ -1,5 +1,8 @@
+#include <chrono>
```
Comment (stuqdog) [inline src/viam/sdk/components/camera/server.cpp:1]: See above comment re: moving this block down.

File: src/viam/sdk/components/camera/server.cpp [hunk 4]
```diff
@@ -60,11 +63,63 @@ ::grpc::Status CameraServer::GetImage(::grpc::ServerContext* context,
...
+    }
+
+    const std::shared_ptr<Camera> camera = std::dynamic_pointer_cast<Camera>(rb);
+
+    const Camera::image_collection image_coll = camera->get_images();
+    for (const auto& img : image_coll.images) {
+        ::viam::component::camera::v1::Image proto_image;
+        const std::string img_string = bytes_to_string(img.bytes);
+        proto_image.set_source_name(img.source_name);
+        proto_image.set_format(MIME_string_to_format(img.mime_type));
```
Comment (stuqdog) [inline src/viam/sdk/components/camera/server.cpp:114]: (q) do you think it makes sense to do this MIME type normalization for `get_image` code as well, as a flyby?

**Review by stuqdog — CHANGES_REQUESTED** [review]

File: src/viam/sdk/components/camera/camera.cpp [hunk 5]
```diff
@@ -66,9 +69,54 @@ Camera::raw_image Camera::from_proto(viam::component::camera::v1::GetImageRespon
...
+    std::chrono::seconds seconds(timestamp.seconds());
+    std::chrono::nanoseconds nanos(timestamp.nanos());
+    return std::chrono::system_clock::time_point(
+        std::chrono::duration_cast<std::chrono::system_clock::duration>(seconds) + nanos);
+}
+
+Camera::image_collection Camera::from_proto(viam::component::camera::v1::GetImagesResponse proto) {
+    Camera::image_collection image_collection;
+    std::vector<Camera::raw_image> images;
+    for (const auto& img : proto.images()) {
```
[the remark this answers is not in the document]
Comment (bhaney) [reply src/viam/sdk/components/camera/camera.cpp:106]: The `GetImageResponse` and `GetImagesResponse` look pretty different in the API: For one, `GetImagesResponse` returns an `Image` message, while `GetImage` does not (it just exposes certain fields). A `GetImagesResponse` does not contain a `GetImageResponse` within it

File: src/viam/sdk/components/camera/camera.cpp [hunk 6]
```diff
@@ -66,9 +69,54 @@ Camera::raw_image Camera::from_proto(viam::component::camera::v1::GetImageRespon
...
+        case viam::component::camera::v1::FORMAT_RAW_RGBA:
+            return "image/vnd.viam.rgba";
+        case viam::component::camera::v1::FORMAT_RAW_DEPTH:
+            return "image/vnd.viam.dep";
+        case viam::component::camera::v1::FORMAT_JPEG:
+            return "image/jpeg";
+        case viam::component::camera::v1::FORMAT_PNG:
+            return "image/png";
+        case viam::component::camera::v1::FORMAT_UNSPECIFIED:
+            return "";
```
[the remark this answers is not in the document]
Comment (bhaney) [reply src/viam/sdk/components/camera/camera.cpp:88]: done

File: src/viam/sdk/components/camera/server.cpp (hunk 3 above)
Comment (bhaney) [reply src/viam/sdk/components/camera/server.cpp:1]: moved to camera.hpp

File: src/viam/sdk/components/camera/server.cpp [hunk 7]
```diff
@@ -60,11 +63,63 @@ ::grpc::Status CameraServer::GetImage(::grpc::ServerContext* context,
     return ::grpc::Status();
 }
 
+// helper function to replace a MIME string with a protobuf format enum
+::viam::component::camera::v1::Format MIME_string_to_format(std::string mime_string) {
```
[the remark this answers is not in the document]
Comment (bhaney) [reply src/viam/sdk/components/camera/server.cpp:67]: put these helper functions in camera.cpp and delcared them in the camera.hpp header as static functions

**Pull request merged**